### DISCUSSION TITLE
Handling Train/Val/Test splits for a small pool of videos with static backgrounds

### DISCUSSION THREAD

**Discussion by cosminlovin7** [ultralytics/ultralytics#23439] [Q&A] [answered]
I am building an object detection model using a small pool of high-resolution videos (approx. 10 videos, with resolution of 3840 x 2160). A key characteristic of this data is that the background location is static, basically all videos are filmed in the same location. The only thing that differs is the weather and time of the day.

Currently, I am weighing two conflicting strategies for dataset splitting:

1. Splitting by Video (Group Splitting): Assigning 7 videos to Train, 2 to Validation, and 1 to Test. The main advantage is that this strategy ensures zero temporal data leakage and proves a fair evaluation of the model. However, with only 10 videos, the model misses out on the specific object variances (lighting, angles, poses) present in the 3 excluded videos during training, which will be part of test and valid splits.

2. Splitting by Frame (Random/Shuffle Splitting): Shuffling all frames from all 10 videos together and splitting 80/10/10. The advantage in this situation would be that the model learns from the full diversity of all 10 videos. However, we have data leakage as the background is static and frames are temporally linked, the validation and test set will lead to inflated mAP metrics.

I am asking this because I have noticed that many datasets on platforms like Kaggle or Roboflow split by frame (second strategy). While this increases the variety of data, it feels like bad practice for production-grade models. However, when the video pool is this small, splitting by video feels like I am starving the model of valuable training features.

I want to know your opinion. In scenarios with very few videos (e.g., <15), do you still strictly adhere to Splitting by Video (first strategy)? Are there "middle-ground" strategies? For example, taking the first 70% of every video for training and the last 30% for validation to maintain some temporal separation while keeping scenes from all videos in the training set? How do you handle the trade-off between feature diversity (seeing all videos) and validation integrity (preventing leakage) in static-background environments with a small pool of videos? I am interested in what are the industry best practices in such situations.

Thank you!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @cosminlovin7, thank you for your interest in Ultralytics 🚀! We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

This is an automated response 🤖—an Ultralytics engineer will also assist soon 😊.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo26"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Reply by glenn-jocher** [reply, marked as answer]
With a static camera/background, shuffling frames across videos will almost always leak scene cues and inflate mAP, so for any “honest” generalization estimate I’d keep validation/test split **by video (or by day/session)**; when you only have ~10 videos, the usual middle-ground is to keep a small **held-out video test set** (never touched) and do **grouped K-fold/leave-one-video-out cross-validation** on the remaining videos for model selection, as shown in our [K-Fold cross-validation guide](https://docs.ultralytics.com/guides/kfold-cross-validation/) (swap `KFold` for `GroupKFold` with `groups=video_id`). If your production setting truly is the same fixed location/camera, a frame-level split can still be useful as an “in-distribution regression test,” but treat it explicitly as such and rely on the video-level split to avoid the data leakage pitfalls described in [the model testing guide](https://docs.ultralytics.com/guides/model-testing/).

**Reply by cosminlovin7** [reply]
That's an interesting suggestion. Thanks for the info!

I also have another scenario now looking for a best practice method (if any). Considering that I want to re-use an existing dataset, which has a high data leakage, with frames from same video splitted between train, test and valid and that there is no way of identifying the frames that may be part of the same video (for example, by looking at the name and identifying a similar prefix), is there a method to somehow restructure the entire dataset to reduce the data leakage using only the data in that dataset?

I know that one way would be to just put the entire data in the training set and only rebuild the test and valid splits with data from another source but i wonder if there is a solution using only the data available in that set. Thank you again!

**Reply by glenn-jocher** [reply]
If you’ve lost any notion of “which frames came from which video/session”, you can’t fully eliminate leakage using only that dataset; the best you can do is rebuild the split by *grouping visually-similar frames* (near-duplicates / same background) and then splitting **by group** so similar frames never land in both train and val/test—then write new `autosplit_*.txt` files (same format Ultralytics uses; see the [data split utilities reference](https://docs.ultralytics.com/reference/data/split/)).

```python
from pathlib import Path
from PIL import Image
import imagehash
from sklearn.model_selection import GroupShuffleSplit

images_dir = Path("path/to/dataset/images")
imgs = sorted([p for p in images_dir.rglob("*.*") if p.suffix.lower() in {".jpg", ".jpeg", ".png"}])

def phash(p: Path) -> str:
    with Image.open(p) as im:
        return str(imagehash.phash(im))

groups = [phash(p) for p in imgs]  # similar background/frames => same (or close) hash

gss = GroupShuffleSplit(n_splits=1, test_size=0.2, random_state=0)
train_idx, tmp_idx = next(gss.split(imgs, groups=groups))

gss2 = GroupShuffleSplit(n_splits=1, test_size=0.5, random_state=0)
val_rel, test_rel = next(gss2.split([imgs[i] for i in tmp_idx], groups=[groups[i] for i in tmp_idx]))
val_idx = [tmp_idx[i] for i in val_rel]
test_idx = [tmp_idx[i] for i in test_rel]

root = images_dir.parent  # where autosplit_*.txt will live
def write_txt(name, idxs):
    (root / name).write_text("".join(f"./images/{imgs[i].name}\n" for i in idxs))

write_txt("autosplit_train.txt", train_idx)
write_txt("autosplit_val.txt", val_idx)
write_txt("autosplit_test.txt", test_idx)
```

This won’t be as “clean” as true video-level grouping, but it usually beats a naive frame shuffle when the background is static; if you can source *any* small external test set later, that’s still the only reliable way to validate real-world generalization.

**Reply by cosminlovin7** [reply]
Thank you very much for your guidance!

**Reply by glenn-jocher** [reply]
Happy to help—if you hit any snags implementing the grouping-based re-split, paste the full traceback (as text) plus your `yolo checks` output here and we can take a look.
